### PR TITLE
Create tensor grids with same rank as device grid bug fix #500

### LINKED ISSUE #500: MNIST inference crashes
When running inference using a cleaner MNIST model:

    def __init__(self, input_size=784, output_size=10, hidden_size=256):
        super(MNISTLinear, self).__init__()
        self.l1 = nn.Linear(input_size, hidden_size)
        self.relu = nn.ReLU()
        self.l2 = nn.Linear(hidden_size, output_size)

    def forward(self, x):
        x = self.l1(x)
        x = self.relu(x)
        x = self.l2(x)
        return nn.functional.softmax(x)

We get the following MLIR error:

2024-08-26 16:23:01.436 | INFO | MLIRCompiler - MLIR module generated successfully.
module @"tt-forge-graph" attributes {tt.system_desc = #tt.system_desc<[{arch = <wormhole_b0>, grid = 8x8, l1_size = 1499136, num_dram_channels = 12, dram_channel_size = 1073741824, noc_l1_address_align_bytes = 16, pcie_address_align_bytes = 32, noc_dram_address_align_bytes = 32, physical_cores = {worker = [(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (1, 0), (1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (2, 0), (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (3, 0), (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6), (3, 7), (4, 0), (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6), (4, 7), (5, 0), (5, 1), (5, 2), (5, 3), (5, 4), (5, 5), (5, 6), (5, 7), (6, 0), (6, 1), (6, 2), (6, 3), (6, 4), (6, 5), (6, 6), (6, 7), (7, 0), (7, 1), (7, 2), (7, 3), (7, 4), (7, 5), (7, 6), (7, 7)] dram = [(8, 0), (9, 0), (10, 0), (8, 1), (9, 1), (10, 1), (8, 2), (9, 2), (10, 2), (8, 3), (9, 3), (10, 3)]}}], [0], [3 : i32], [<0, 0, 0, 0>]>} {
func.func https://github.com/forward(%arg0: tensor<1x784xf32> {ttir.name = "input_1"}, %arg1: tensor<10xf32> {ttir.name = "l2.bias"}, %arg2: tensor<256x10xf32> {ttir.name = "l2.weight"}, %arg3: tensor<256xf32> {ttir.name = "l1.bias"}, %arg4: tensor<784x256xf32> {ttir.name = "l1.weight"}) -> tensor<1x10xf32> {
%0 = tensor.empty() : tensor<1x256xf32>
%1 = "ttir.matmul"(%arg0, %arg4, %0) <{operand_constraints = [#tt.operand_constraint<dram|l1|scalar|tile|any_device|any_device_tile>, #tt.operand_constraint<dram|l1|scalar|tile|any_device|any_device_tile>, #tt.operand_constraint<dram|l1|scalar|tile|any_device|any_device_tile>]}> : (tensor<1x784xf32>, tensor<784x256xf32>, tensor<1x256xf32>) -> tensor<1x256xf32>
%2 = tensor.empty() : tensor<1x256xf32>
%3 = "ttir.add"(%1, %arg3, %2) <{operandSegmentSizes = array<i32: 2, 1>, operand_constraints = [#tt.operand_constraint<dram|l1|scalar|tile|any_device|any_device_tile>, #tt.operand_constraint<dram|l1|scalar|tile|any_device|any_device_tile>, #tt.operand_constraint<dram|l1|scalar|tile|any_device|any_device_tile>]}> : (tensor<1x256xf32>, tensor<256xf32>, tensor<1x256xf32>) -> tensor<1x256xf32>
%4 = tensor.empty() : tensor<1x256xf32>
%5 = "ttir.relu"(%3, %4) <{operandSegmentSizes = array<i32: 1, 1>, operand_constraints = [#tt.operand_constraint<dram|l1|scalar|tile|any_device|any_device_tile>, #tt.operand_constraint<dram|l1|scalar|tile|any_device|any_device_tile>]}> : (tensor<1x256xf32>, tensor<1x256xf32>) -> tensor<1x256xf32>
%6 = tensor.empty() : tensor<1x10xf32>
%7 = "ttir.matmul"(%5, %arg2, %6) <{operand_constraints = [#tt.operand_constraint<dram|l1|scalar|tile|any_device|any_device_tile>, #tt.operand_constraint<dram|l1|scalar|tile|any_device|any_device_tile>, #tt.operand_constraint<dram|l1|scalar|tile|any_device|any_device_tile>]}> : (tensor<1x256xf32>, tensor<256x10xf32>, tensor<1x10xf32>) -> tensor<1x10xf32>
%8 = tensor.empty() : tensor<1x10xf32>
%9 = "ttir.add"(%7, %arg1, %8) <{operandSegmentSizes = array<i32: 2, 1>, operand_constraints = [#tt.operand_constraint<dram|l1|scalar|tile|any_device|any_device_tile>, #tt.operand_constraint<dram|l1|scalar|tile|any_device|any_device_tile>, #tt.operand_constraint<dram|l1|scalar|tile|any_device|any_device_tile>]}> : (tensor<1x10xf32>, tensor<10xf32>, tensor<1x10xf32>) -> tensor<1x10xf32>
%10 = tensor.empty() : tensor<1x10xf32>
%11 = "ttir.softmax"(%9, %10) <{dimension = 1 : si32, operand_constraints = [#tt.operand_constraint<dram|l1|scalar|tile|any_device|any_device_tile>, #tt.operand_constraint<dram|l1|scalar|tile|any_device|any_device_tile>]}> : (tensor<1x10xf32>, tensor<1x10xf32>) -> tensor<1x10xf32>
return %11 : tensor<1x10xf32>
}
}
2024-08-26 16:23:01.437 | INFO | MLIRCompiler - MLIR module generated successfully.
2024-08-26 16:23:01.438 | INFO | MLIRCompiler - MLIR passes run successfully.
python: /opt/ttmlir-toolchain/src/llvm-project/mlir/lib/IR/AffineMap.cpp:544: mlir::AffineMap mlir::AffineMap::compose(mlir::AffineMap) const: Assertion `getNumDims() == map.getNumResults() && "Number of results mismatch"' failed.
Fatal Python error: Aborted

### PR DESCRIPTION
**Pull request by nsmithtt**
Previously we were just creating tensor layout grids with rank 1 or 2 depending on the tensor's rank.  This however is incorrect, the tensor grid must be of the same rank as the device grid.

The fix is to use the device grid's rank in the layout type converter to ensure that by default a tensor layout gets a grid of equivalent rank.

### CODE REVIEW

**Review by uazizTT — APPROVED** [review]

**Review by sdjordjevicTT — APPROVED** [review]
Great change Nick! Thanks for the quick fix!
